runtime/serde/serde-xml/common/src/aws/smithy/kotlin/runtime/serde/xml/XmlDeserializer.kt

```diff
@@ -213,65 +214,79 @@ internal class XmlListDeserializer(
  * @param parentToken initial token of associated structure
  * @param parsedFieldLocations list of [FieldLocation] representing values able to be loaded into deserialized instances
  */
-internal class XmlStructDeserializer(
+private class XmlStructDeserializer(
     private val objDescriptor: SdkObjectDescriptor,
-    private val reader: XmlStreamReader,
+    private var reader: XmlStreamReader,
     private val parentToken: XmlToken.BeginElement,
     private val parsedFieldLocations: MutableList<FieldLocation> = mutableListOf(),
+    private val unwrapped: Boolean,
+    private var firstIteration: Boolean = true,
 ) : Deserializer.FieldIterator {
     // Used to track direct deserialization or further nesting between calls to findNextFieldIndex() and deserialize<Type>()
     private var reentryFlag: Boolean = false
 
     override fun findNextFieldIndex(): Int? {
-        if (inNestedMode()) {
-            // Returning from a nested struct call.  Nested deserializer consumed
-            // tokens so clear them here to avoid processing stale state
-            parsedFieldLocations.clear()
-        }
+        if (unwrapped) {
+            val FIELD_INDEX: Int = 0
+            return if (reader.peek() is XmlToken.Text) FIELD_INDEX else null
+        } else {
+            if (firstIteration) reader = reader.subTreeReader(XmlStreamReader.SubtreeStartDepth.CHILD); firstIteration = false
```

Comment:
**Question**: Using a mutable `reader` and changing its value on the first iteration seems clunky to me. Could detecting the wrapping mode be done during class initialization?

```kotlin
private class XmlStructDeserializer(
    ...
    reader: XmlStreamReader, // **not** a private val, just a constructor param
    ...
    unwrapped: Boolean, // **not** a private val, just a constructor param
) : Deserializer.FieldIterator {
    private val reader = if (unwrapped) reader else reader.subTreeReader()
```